### ISSUE TITLE
Converge fails copying files into instance

### ISSUE THREAD

**Issue by benjamink** [neillturner/kitchen-ansible#173]
I've just recently started getting this error when doing a `kitchen test`:

```
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #converge action: [No such file or directory @ rb_file_s_stat - /var/folders/lt/83z_f5k17_g68tzghy47q02h0000gn/T/pm-inbound-centos-67-sandbox-20160601-61944-gg84p3/roles/newrelic-meetme-plugin/.git/config]
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

The directory `/var/folders/lt/83z_f5k17_g68tzghy47q02h0000gn/T/pm-inbound-centos-67-sandbox-20160601-61944-gg84p3` doesn't exist.

I tried downgrading both `test-kitchen` to 1.8.0 & `kitchen-ansible` to 0.42.1 but it didn't make a difference.  Not sure where to start debugging.

```
$ vagrant --version
Vagrant 1.8.1

$ VBoxManage --version
5.0.20r106931
```

Running on OS X 10.11.5

Kitchen log:

```
I, [2016-06-01T20:37:56.434486 #61944]  INFO -- Kitchen: -----> Starting Kitchen (v1.8.0)
I, [2016-06-01T20:37:56.821125 #61944]  INFO -- Kitchen: -----> Converging <pm-inbound-centos-67>...
E, [2016-06-01T20:37:57.536459 #61944] ERROR -- Kitchen: ------Exception-------
E, [2016-06-01T20:37:57.536502 #61944] ERROR -- Kitchen: Class: Kitchen::ActionFailed
E, [2016-06-01T20:37:57.536517 #61944] ERROR -- Kitchen: Message: Failed to complete #converge action: [No such file or directory @ rb_file_s_stat - /var/folders/lt/83z_f5k17_g68tzghy47q02h0000gn/T/pm-inbound-centos-67-sandbox-20160601-61944-gg84p3/roles/newrelic-meetme-plugin/.git/config]
E, [2016-06-01T20:37:57.536555 #61944] ERROR -- Kitchen: ---Nested Exception---
E, [2016-06-01T20:37:57.536615 #61944] ERROR -- Kitchen: Class: Errno::ENOENT
E, [2016-06-01T20:37:57.536630 #61944] ERROR -- Kitchen: Message: No such file or directory @ rb_file_s_stat - /var/folders/lt/83z_f5k17_g68tzghy47q02h0000gn/T/pm-inbound-centos-67-sandbox-20160601-61944-gg84p3/roles/newrelic-meetme-plugin/.git/config
E, [2016-06-01T20:37:57.536651 #61944] ERROR -- Kitchen: ------Backtrace-------
E, [2016-06-01T20:37:57.536665 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp/upload.rb:117:in `stat'
E, [2016-06-01T20:37:57.536678 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp/upload.rb:117:in `set_current'
E, [2016-06-01T20:37:57.536689 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp/upload.rb:103:in `next_item_state'
E, [2016-06-01T20:37:57.536700 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp.rb:400:in `await_response_state'
E, [2016-06-01T20:37:57.536711 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp.rb:369:in `block (3 levels) in start_command'
E, [2016-06-01T20:37:57.536722 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/channel.rb:311:in `process'
E, [2016-06-01T20:37:57.536732 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:223:in `block in preprocess'
E, [2016-06-01T20:37:57.536750 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:223:in `each'
E, [2016-06-01T20:37:57.536761 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:223:in `preprocess'
E, [2016-06-01T20:37:57.536773 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:206:in `process'
E, [2016-06-01T20:37:57.536783 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:170:in `block in loop'
E, [2016-06-01T20:37:57.536794 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:170:in `loop'
E, [2016-06-01T20:37:57.536805 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/session.rb:170:in `loop'
E, [2016-06-01T20:37:57.536815 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-ssh-3.0.1/lib/net/ssh/connection/channel.rb:269:in `wait'
E, [2016-06-01T20:37:57.536826 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/net-scp-1.2.1/lib/net/scp.rb:284:in `upload!'
E, [2016-06-01T20:37:57.536837 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:151:in `block in upload'
E, [2016-06-01T20:37:57.536857 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:148:in `each'
E, [2016-06-01T20:37:57.536872 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:148:in `upload'
E, [2016-06-01T20:37:57.536883 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/provisioner/base.rb:70:in `block in call'
E, [2016-06-01T20:37:57.536894 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/base.rb:86:in `initialize'
E, [2016-06-01T20:37:57.536904 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:350:in `new'
E, [2016-06-01T20:37:57.536915 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:350:in `create_new_connection'
E, [2016-06-01T20:37:57.536926 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/transport/ssh.rb:85:in `connection'
E, [2016-06-01T20:37:57.536936 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/provisioner/base.rb:66:in `call'
E, [2016-06-01T20:37:57.536946 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:373:in `block in converge_action'
E, [2016-06-01T20:37:57.536957 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:513:in `synchronize_or_call'
E, [2016-06-01T20:37:57.536968 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:478:in `block in action'
E, [2016-06-01T20:37:57.536982 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/2.4.0/benchmark.rb:293:in `measure'
E, [2016-06-01T20:37:57.536993 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:477:in `action'
E, [2016-06-01T20:37:57.537004 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:369:in `converge_action'
E, [2016-06-01T20:37:57.537014 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:348:in `block in transition_to'
E, [2016-06-01T20:37:57.537025 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:347:in `each'
E, [2016-06-01T20:37:57.537035 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:347:in `transition_to'
E, [2016-06-01T20:37:57.537046 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/instance.rb:138:in `converge'
E, [2016-06-01T20:37:57.537056 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/command.rb:176:in `public_send'
E, [2016-06-01T20:37:57.537067 #61944] ERROR -- Kitchen: /Users/bkrein/.rbenv/versions/2.3.0-dev/lib/ruby/gems/2.4.0/gems/test-kitchen-1.8.0/lib/kitchen/command.rb:176:in `block (2 levels) in run_action'
E, [2016-06-01T20:37:57.537090 #61944] ERROR -- Kitchen: ----------------------
```

**Comment by neillturner** [comment]
well first  run 
   kitchen create  xxxx  -l debug 
and check this is ok. it should be as you are failing in converge
 then run 
   kitchen converge  xxxx  -l debug 
and see what log says
  the provisioner calls the test-kitchen core to create a sandbox and copy files into it. 
post logs and your kitchen.yml file and we have a look/

**Comment by benjamink** [comment]
Create succeeds:
[pm-inbound-centos-67_create.txt](https://github.com/neillturner/kitchen-ansible/files/295547/pm-inbound-centos-67_create.txt)

Converge seems to create the copy directory then delete it before using it:
[pm-inbound-centos-67_converge.txt](https://github.com/neillturner/kitchen-ansible/files/295549/pm-inbound-centos-67_converge.txt)

`.kitchen.yml` renamed to `kitchen.txt` for attaching:
[kitchen.txt](https://github.com/neillturner/kitchen-ansible/files/295551/kitchen.txt)

**Comment by neillturner** [comment]
Well it is normal that test-kitchen deletes the sand-box when it finishes. 
I don't think it's a good idea to set 
      additional_copy_path:
        - "."
try removing this and testing again. 
if that "works" then you need to re-think thid additonal_copy_pathh requirement

**Comment by benjamink** [comment]
The additional_copy_path doesn't seem to make a difference.  Removing it doesn't change the outcome at all.  

[pm-inbound-centos-67_create_nocopypath.txt](https://github.com/neillturner/kitchen-ansible/files/295661/pm-inbound-centos-67_create_nocopypath.txt)

[pm-inbound-centos-67_converge_nocopypath.txt](https://github.com/neillturner/kitchen-ansible/files/295662/pm-inbound-centos-67_converge_nocopypath.txt)

[kitchen_nocopypath.txt](https://github.com/neillturner/kitchen-ansible/files/295665/kitchen_nocopypath.txt)

**Comment by neillturner** [comment]
Well the problem is file: 
  /var/folders/lt/83z_f5k17_g68tzghy47q02h0000gn/T/pm-inbound-centos-67-sandbox-20160602-3490-gy9r5o/roles/newrelic-meetme-plugin/.git/config

I would try removing the /newrelic-meetme-plugin  role as see if it works. (well copies correctly) 
Also you can terminate kitchen converge after it has created the same box by CTL-D and then look at the sandbox.  
Or getting the roles are your using ansible via requirements requirements-test.txt. Maybe you have not got the ansible galaxy access setup. try testting without the
 requirements_path: requirements-test.txt
and see if the failure is then somewhere else like when you run.

**Comment by benjamink** [comment]
I think I might have found the problem.  Need to do some testing.  We recently switched to using Google's `git-repo` tool to manage incorporating our role repos into the main Ansible repository but `git-repo` creates symlinks in all the role's `.git/` directories:

```
[git] $ ls -la roles/newrelic-meetme-plugin/.git
total 104
drwxr-xr-x  15 bkrein  staff   510 Jun  1 11:34 .
drwxr-xr-x  20 bkrein  staff   680 Jun  1 11:34 ..
-rw-r--r--   1 bkrein  staff    41 Jun  1 11:34 HEAD
lrwxr-xr-x   1 bkrein  staff    63 Jun  1 11:34 config -> ../../../.repo/projects/roles/newrelic-meetme-plugin.git/config
lrwxr-xr-x   1 bkrein  staff    73 Jun  1 11:34 description -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/description
lrwxr-xr-x   1 bkrein  staff    67 Jun  1 11:34 hooks -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/hooks
-rw-r--r--   1 bkrein  staff  2830 Jun  1 11:34 index
lrwxr-xr-x   1 bkrein  staff    66 Jun  1 11:34 info -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/info
lrwxr-xr-x   1 bkrein  staff    61 Jun  1 11:34 logs -> ../../../.repo/projects/roles/newrelic-meetme-plugin.git/logs
lrwxr-xr-x   1 bkrein  staff    69 Jun  1 11:34 objects -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/objects
lrwxr-xr-x   1 bkrein  staff    68 Jun  1 11:34 packed-refs -> ../../../.repo/projects/roles/newrelic-meetme-plugin.git/packed-refs
lrwxr-xr-x   1 bkrein  staff    61 Jun  1 11:34 refs -> ../../../.repo/projects/roles/newrelic-meetme-plugin.git/refs
lrwxr-xr-x   1 bkrein  staff    70 Jun  1 11:34 rr-cache -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/rr-cache
lrwxr-xr-x   1 bkrein  staff    64 Jun  1 11:34 shallow -> ../../../.repo/projects/roles/newrelic-meetme-plugin.git/shallow
lrwxr-xr-x   1 bkrein  staff    65 Jun  1 11:34 svn -> ../../../.repo/project-objects/newrelic-meetme-plugin.git.git/svn
```

I suspect this is what's breaking test-kitchen.  I'll do some re-arranging of things & see what happens.

**Comment by neillturner** [comment]
Oh yes symlinks are bad it does handle very well at all. not sure how to fix though. 
on kitchen-puppet i added a parameter: 
  ignored_paths_from_root: 
See: https://github.com/neillturner/kitchen-puppet/commit/00f8b9f1f37e2d9ad45bfb6dba2a60675614275d
we could do something similar.

**Comment by benjamink** [comment]
Can that be setup to ignore `.git/` paths in general?

**Comment by benjamink** [comment]
One thing I noticed that complicates this process is that the `requirements_path` option in kitchen-ansible requires a file formatted differently than what ansible-galaxy can use.  For instance, it doesn't allow for using Git references to internally hosted roles.  For the use-case I have here where we're trying to do integrations of multiple roles to build a system that'd be really helpful.  I assume we'd need to be able to pass a private key or something in that is able to read from internal private Git repos.

**Comment by neillturner** [comment]
everything is possible. you can code your own ansible-playbook custom command and put the requirements parameter as you want it and leave the requirements_path parameter empty in kitchen.yml. 
or use the ansible_extra_flags  to add your requirements_path parameter.

**Comment by benjamink** [comment]
I came up with a really crappy fix but I'm not a Ruby dev so I'd love to know if there's a more elegant solution.  Basically the problem is that `FileUtils.cp_r` won't copy symlinks & there's no way to dereference them that I could find.  I don't expect this to be a workable solution so I'm not submitting it as a PR, but here's the change I made: 

https://github.com/benjamink/kitchen-ansible/commit/9205273ad77a74046ab9a7b32b8c2b58bacc8f5d

Note that this doesn't cover all the directories that I'd envision having a problem with potentially, just the `roles/` directory really.

**Comment by neillturner** [comment]
if you want to run operating system commands that you need to wrap them in a system or sheelout ruby statement but let look into FileUils for issues about symlink.

**Comment by benjamink** [comment]
I had read about that option, but unfortunately adding the `{remove_destination: true}` didn't help:

[pm-inbound-centos-67_create_removedest.txt](https://github.com/neillturner/kitchen-ansible/files/302581/pm-inbound-centos-67_create_removedest.txt)

[pm-inbound-centos-67_converge_removedest.txt](https://github.com/neillturner/kitchen-ansible/files/302582/pm-inbound-centos-67_converge_removedest.txt)

[ansible_playbook-rmdest.diff.txt](https://github.com/neillturner/kitchen-ansible/files/302584/ansible_playbook-rmdest.diff.txt)

I'll try via Ruby's `system` statement.

**Comment by neillturner** [comment]
also try this:  
FileUtils.cp_r(file, destination,{:verbose => true, :remove_destination => true} )

**Comment by neillturner** [comment]
copy use this routine:  
https://gist.github.com/akostadinov/fc688feba7669a4eb784 to implement cp_r_dereference.rb
and its based on this code: 
https://gist.github.com/akostadinov/05c2a976dc16ffee9cac
in ruby you can add your own methods to any object. 
if you have a test case that i can reproduce then i can have a go at a fix.

**Comment by benjamink** [comment]
The `system()` copy command didn't work, but I need to debug why.  I did the exact same thing with `system()` that worked with backticks.

In the meantime, I setup a demo Ansible directory configured for `kitchen-ansible` & `git-repo` here: https://github.com/benjamink/ansible-git-repo

The `repo` command listed there will pull in the example `helloworld` role I created here: https://github.com/benjamink/ansible-helloworld

I wrote instructions in the README that should let you duplicate the issues I'm having.

**Comment by benjamink** [comment]
The `system()` function works fine, I must have been editing the wrong version of `kitchen-ansible`:

```diff
@@ -820,7 +820,7 @@
         # If so, make sure to copy into VM so dir structure is like: /tmp/kitchen/roles/role_name

         FileUtils.mkdir_p(File.join(tmp_roles_dir, role_name))
-        FileUtils.cp_r(Dir.glob("#{roles}/*"), File.join(tmp_roles_dir, role_name))
+        system("cp -LR #{roles} #{File.join(tmp_roles_dir, role_name)}")
       end

       # copy ansible.cfg if found in root of repo
@@ -893,7 +893,7 @@
           destination = File.join(sandbox_path, File.basename(file))
           if File.directory?(file)
             info("Copy dir: #{file} #{destination}")
-            FileUtils.cp_r(file, destination)
+            system("cp -LR #{file} #{destination}")
           else
             info("Copy file: #{file} #{destination}")
             FileUtils.cp file, destination
```

I think with the `copy_symlinks: false` option you mentioned to toggle the above, that'd work nicely.  I imagine the toggle should be added to the other functions that do recursive copies as well?

**Comment by neillturner** [comment]
ok i will add a copy_simlink option that defaults to false and when true will do the cp command. We already have have remove_destination: true on a number of cp_r commands so i will add t the rest,

**Comment by neillturner** [comment]
closed fixed in 0.44.1 by adding  ignore_paths_from_root option to handle symlinks in copied directories #182
